use new health service with rolling updates

esuen opened the pull request:
Changes required for rolling updates to use the health service client for health checks while still using the consul health tree in parallel. These changes will test reliability of health service client.
- Created a separate healthChecker type in checker.go called ShadowTrafficHealthChecker that will utilize both the health service client and old consul tree checking code
- Old health checker was renamed from ConsulHealthChecker to HealthChecker so that in the future when the health service client is deemed reliable we can rename ShadowTrafficHealthChecker to HealthChecker
- WatchService and Service functions defined for new HealthChecker to be used by rolling updates
- WatchService uses both health service client and consul health tree
- Service uses either consul or health service client

rudle commented on pkg/health/checker/checker.go:
```diff
@@ -5,14 +5,171 @@ import (
+	podID types.PodID,
+	quitCh <-chan struct{},
+) (chan health.Result, chan error) {
+	return consulHealthChecker(h.(healthServiceChecker)).WatchPodOnNode(nodename, podID, quitCh)
```
What's the long term fate for this type (`consulHealthChecker`)? What will happen when we move to only health checking via health service? I predicted that this code would introduce a new type that "covered" both consulHealthChecker and healthServiceHealthChecker and we'd use that everywhere. Can you remind me why that's not possible?

esuen replied:
That's probably the better approach in the long term. That way when consulHealthChecker code can be removed when we move to only the health service. I'll make those changes.

esuen commented:
Ready for review, there's some minor TODOs I need to look at and some testing to be done.

esuen commented:
fixing checker, don't review

esuen commented:
you may review again..

rudle commented on pkg/health/checker/checker.go:
```diff
@@ -41,28 +69,37 @@ type healthStore interface {
-		client:      client,
-		kv:          client.KV(),
-		consulStore: consul.NewConsulStore(client),
+func NewHealthChecker(hClient HealthClient, resourceClient ResourceClient, cClient consulutil.ConsulClient, rcStore ReplicationControllerStore, labelReader LabelReader) HealthChecker {
```
WDYT about a new constructor for clients who want to specify all of these arguments? The behaviour is special, so a special constructor isn't unreasonable. How about NewShadowTrafficHealthChecker(…). It has the benefit of being a bit clearer about intent and doesn't require the caller to specify so many `nil`s. At the very least, this constructor should declare which arguments are optional and which aren't.

esuen replied:
Are you suggesting to rename NewHealthChecker to NewShadowTrafficHealthChecker with optional parameters or to create an additional constructor.

Would it be implicitly or explicitly specified that certain parameters are optional? Golang doesn't seem to have a way to specify optional parameters unless the function takes a struct and the fields in the struct are optional I think.

mpuncel replied:
I think he's suggesting a totally separate constructor, like `NewShadowTrafficHealthChecker` and leave the old `NewHealthChecker` alone for now. Then when we're confident in the health API service we can do a rename and drop the old consul related arguments

rudle replied:
^ yeah, this sounds right. It's easier to find callers of the old constructor than it is to find callers of a shared constructor that use deprecated arguments.

rudle replied:
In all the call sites where there are several nils passed in to the constructor, I suggest using a constructor that doesn't require the caller to pass in those nils. The NewShadowTrafficHealthChecker can call into the main constructor and pass in the nil arguments itself. This is a style request, it doesn't have much bearing on correctness.

Here's an example: https://play.golang.org/p/SaCDXIJO13-

esuen replied:
Cool, thanks. Why the name "ShadowTraffic"?

mpuncel replied:
"shadowing" traffic is the technique of duplicating all of the traffic going to destination A and sending it also to destination B. Usually this is done to see if destination B can handle the load that destination A can, and it's done when you are evaluating replacing A with B.

That's basically what we're doing here, we want to see if the health service can handle being accessed every time we used to be looking in consul

rudle commented on pkg/health/checker/checker.go:
```diff
@@ -213,75 +278,174 @@ func (c consulHealthChecker) WatchHealth(
-				case <-quitCh:
-					return
-				case errCh <- consulutil.NewKVError("list", consul.HealthPath(serviceID, "/"), err):
+	if !useHealthService {
```
Is there any way to get this to send health check requests to both consul & the health service? I don't yet have enough confidence that the health service returns timely & correct health checks, which is why I want this code to send traffic to the trusted source (_consul_) and the new service so we can see how it behaves under production load levels.

esuen replied:
Would this always send to both or only when useHealthService is true?

rudle replied:
We can use the boolean to toggle whether or not this code sends to both.

rudle commented on pkg/health/checker/checker.go:
```diff
@@ -265,23 +345,110 @@ func (c consulHealthChecker) WatchService(
+					continue
+				}
+				for _, endpoint := range statusEndpoints {
+					healthReq := &hclient.HealthRequest{
```
Does this need to change given the new health service proto? It's kind of annoying that the protos are closed source - it creates a bit of a circular dependency.

esuen replied:
This doesn't need to change since this is calling HealthMonitor and it still uses HealthRequest messages.

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -41,28 +69,45 @@ type healthStore interface {
 }
 
-func (c consulHealthChecker) WatchPodOnNode(
+func NewShadowTrafficHealthChecker(cClient consulutil.ConsulClient) HealthChecker {
```
This looks great, but you have the constructor names backwards :) this one should be the old untouched `NewConsulHealthChecker` and the other one (which does traffic shadowing to the health service) should be called `NewShadowTrafficHealthChecker`

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -176,10 +221,42 @@ func publishLatestHealth(inCh <-chan api.KVPairs, quitCh <-chan struct{}, result
+	serviceID string,
+	status manifest.StatusStanza,
+) ([]string, error) {
+	rcIDs, err := h.resourceClient.GetRCIDsForPod(types.PodID(serviceID))
```
this function depends on fields that are only set if you called the constructor with more arguments. This means that if you 1) use the old constructor and 2) call `WatchService()` or `Service()` with `useHealthService=true` you're going to get a nil pointer exception!

This is making me think that the shadow checker and the non-shadow checker should probably be different types (not just `healthChecker`) and then their behavior can differ without needing to pass extra arguments like `useHealthChecker` to the functions.

A lot of what they do will be common, you can probably extract those into package level functions pretty easily (I hope). These package level functions would have to take all the struct fields as arguments.

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -176,10 +221,42 @@ func publishLatestHealth(inCh <-chan api.KVPairs, quitCh <-chan struct{}, result
+		statusEndpoints = make([]string, 0, len(labeled))
+		statusPath := status.Path
+		if status.Path == "" {
+			statusPath = "_status"
```
nit: this might be a nice detail to hide in the status stanza struct, e.g. `statusStanza.Path()` and it applies the default if necessary

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -265,23 +345,110 @@ func (c consulHealthChecker) WatchService(
+						Protocol: "HTTPS",
+					}
+
+					err = h.healthClient.HealthMonitor(context.Background(), healthReq, respChan)
```
also the quit channel pattern is a dated go practice, contexts are preferable now

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -213,28 +290,31 @@ func (c consulHealthChecker) WatchHealth(
 	}
 
 	timer := time.NewTimer(0)
+	respChan := make(chan *hclient.HealthResponse, 1)
```
this channel is never closed, which means we're going to leak resources every time `WatchService` is called. We need to make sure that it's closed once we expect no values to be sent on it anymore.

Long winded explanation warning: In my experience (and I think I read in a blog post somewhere), the best way to get channel closing right is to only have the sender on the channel do the closing. In this case, the sender is the `h.healthClient.HealthMonitor()` function, and the receiver is this function (`WatchService()`).

What this means is that I think that instead of declaring `respChan` here in this function, you should have`h.healthClient.HealthMonitor()` create it and return it. Then, `HealthMonitor()` should wait for the passed context to be canceled, and when it is, close the channel and exit the goroutine. This guarantees that no values are sent after the channel is closed (which causes a panic)

esuen replied:
What if there's a 
```golang
defer close(resultCh)
```
in HealthMonitor?

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -265,23 +345,110 @@ func (c consulHealthChecker) WatchService(
+						case resp, ok := <-respChan:
+							if ok {
+								out := make(map[types.NodeName]health.Result)
+								nodeID := types.NodeName(resp.HealthRequest.Url)
```
ah apparently that's going to require some changes to the health client in another repo :(

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -265,23 +345,110 @@ func (c consulHealthChecker) WatchService(
 				case resultCh <- out:
 				}
 			}
+			if useHealthService {
```
you'll also be able to get rid of this `if` if you use different types for the shadow and not-shadow checker

mpuncel commented on bin/p2-watch/main.go:
```diff
@@ -138,12 +140,14 @@ func watchPodClusters(client consulutil.ConsulClient, applicator labels.Applicat
+	ctx, cancelFunc := context.WithCancel(context.Background())
 	watchDelay := 1 * time.Second
-	go hc.WatchService(*healthService, healthResults, errCh, quitCh, watchDelay)
+	go hc.WatchService(*healthService, healthResults, errCh, ctx, watchDelay, false, manifest.StatusStanza{})
```
nit: `ctx` should be the first parameter:

```
The Context should be the first parameter, typically named ctx:
```
https://golang.org/pkg/context/

mpuncel commented on pkg/health/checker/checker.go:
```diff
@@ -41,28 +68,45 @@ type healthStore interface {
-		kv:          client.KV(),
-		consulStore: consul.NewConsulStore(client),
+func NewHealthChecker(cClient consulutil.ConsulClient) HealthChecker {
+	return healthChecker{
```
it still makes me nervous that these two constructors return structs of the same type. If I instantiate a `healthChecker` by calling `NewHealthChecker()`, and then I pass `true` for `useHealthService`, i'm going to end up with a nil pointer exception at runtime. If you use different types to implement the functionality, the "normal" health checker could just ignore that argument

esuen replied:
Thanks for the reminder, I was going to get around to that.

esuen commented:
- added new type for shadowHealthChecker.
- put ctx param as first param in function signatures
- Will be looking into making the grpc call for monitoring multiple endpoints better. Instead of for looping multiple monitor calls, I think it should be one call with all the endpoints

esuen commented:
Will need a final review on this.

esuen commented:
TODO: Remove logging code in checker.go

- Created a separate healthChecker type in checker.go called ShadowTrafficHealthChecker that will utilize both the health service client and old consul tree checking code
- Old health checker was renamed from ConsulHealthChecker to HealthChecker so that in the future when the health service client is deemed reliable we can rename ShadowTrafficHealthChecker to HealthChecker
- Modified references in scripts p2-watch, p2-replicate, p2-rctl, p2-rctl-server, p2-inspect, p2-ds-farm to reference health checker signature or name changes as well as references to WatchService and Service calls
- References for ds/farm.go, ds/daemon_set.go, rc/farm.go, rc/replication_controller.go as well
- References for checker_test.go, fake_checker.go, store_test.go, replication/common_setup_test.go, pkg/roll/update_test.go as well
- References for pkg/health/store/store.go, pkg/replication/health_aggregator.go, pkg/replication/replicator.go, pkg/roll/farm.go as well
- Changes to pkg/manifest.go, add helper functions for checker to access the StatusStanza
- update pkg/roll/run_update.go to use ShadowTrafficHealthChecker

- Majority of changes in checker.go and health app
- New Health Checker type defined
- WatchService and Service functions defined for new HealthChecker to be used by rolling updates
- WatchService uses both health service client and consul health tree
- Make request to health service and wait for responses in a respChan and then pass back to the roll farm
- Service uses either consul or health service client. For health service client its a simple request response and return response

rudle commented on pkg/roll/run_update.go:
```diff
@@ -330,9 +336,13 @@ func (u *update) Run(ctx context.Context) (ret bool) {
 	watchDelay := 1 * time.Second
-	go u.hcheck.WatchService(string(newFields.Manifest.ID()), hChecks, hErrs, hQuit, watchDelay)
+	config := newFields.Manifest.GetConfig()
+	useHealthService, _ := config["use_health_service"].(bool)
```
what if config["use_health_service"] isn't present? That's what the second return value is for.

esuen replied:
useHealthService defaults to false if it's not present.

mpuncel replied:
Are you sure we want this config to be in app pod manifests? it should probably be a setting on the roll farm right?

esuen replied:
I thought we test it with specific apps first but I'm open to making it a global config. What does @rudle think?

rudle replied:
Probably we need both, right? app specific so we can test applications like slug & p2-heartbeat and then roll farm wide so we can turn it on and off with the rest of the apps at Square.

mpuncel replied:
ah okay, I'm just out of the loop

mpuncel replied:
I would still check `ok` to make sure that the field is actually a bool, so you don't crash the server if someone has a typo

esuen replied:
The ok is not used and no panic occurs even if useHealthService is not a bool. See https://tour.golang.org/methods/15, "If not, ok will be false and t will be the zero value of type T, and no panic occurs."

It's only when you do
```golang
useHealthService := config["use_health_service"].(bool)
```
for it to panic when the value is of type not bool.

mpuncel replied:
The question is, what do you want to do if someone _intended_ to set it to true, but messed up up with a typo, e.g. `truee`. You're going to treat it as false. Perhaps that's fine, but it's worth thinking about

mpuncel replied:
I do feel better that it at least won't crash the server, probably for a flag like this it's okay to not introduce errors here which will be difficult to recover from

esuen replied:
I'd think I want only `true` to be treated as true and everything else be false otherwise there needs to be code to interpret what the user implied which is I don't think can be fully correct.

spenceral replied:
You can check out https://golang.org/pkg/strconv/#ParseBool to error on non-booleans

esuen commented:
-Added a field to ShadowTrafficHealthChecker struct useHealthService bool that will take precedence over the useHealthService passed to `Service` and `WatchService` of ShadowTrafficHealthChecker.
- Updated references to ShadowTrafficHealthChecker in `bin/p2-rctl/main.go` and `bin/p2-rctl-server/main.go`

esuen commented:
Close/reopen to rerun travis-ci.

esuen commented:
This can use a quick final review since last time, changes made for the suggestions from last review. See previous comments for details on the changes.
